libraries/Microsoft.Bot.Builder.Runtime/CoreBotAdapter.cs

```diff
@@ -22,17 +24,25 @@ internal class CoreBotAdapter : BotFrameworkHttpAdapter
         /// Initializes a new instance of the <see cref="CoreBotAdapter"/> class.
         /// </summary>
         /// <param name="services">Services registered with the application.</param>
-        /// <param name="configuration">Application configuration.</param>
         /// <param name="options">Configured options for the <see cref="CoreBotAdapter"/> instance.</param>
+        /// <param name="configuration">Application configuration.</param>
+        /// <param name="credentialProvider">The credential provider.</param>
+        /// <param name="authConfig">The authentication configuration.</param>
+        /// <param name="channelProvider">The channel provider.</param>
+        /// <param name="connectorClientRetryPolicy">Retry policy for retrying HTTP operations.</param>
+        /// <param name="customHttpClient">The HTTP client.</param>
+        /// <param name="logger">The ILogger implementation this adapter should use.</param>
         public CoreBotAdapter(
             IServiceProvider services,
+            IOptions<CoreBotAdapterOptions> options,
             IConfiguration configuration,
-            IOptions<CoreBotAdapterOptions> options)
-            : base(
-                services.GetService<ICredentialProvider>(),
-                services.GetService<AuthenticationConfiguration>(),
-                services.GetService<IChannelProvider>(),
-                logger: services.GetService<ILogger<BotFrameworkHttpAdapter>>())
+            ICredentialProvider credentialProvider = null,
+            AuthenticationConfiguration authConfig = null,
+            IChannelProvider channelProvider = null,
+            RetryPolicy connectorClientRetryPolicy = null,
+            HttpClient customHttpClient = null,
+            ILogger logger = null)
+            : base(configuration, credentialProvider: credentialProvider, authConfig: authConfig, channelProvider: channelProvider, connectorClientRetryPolicy: connectorClientRetryPolicy, customHttpClient: customHttpClient, logger: logger)
```

Comment:
I'm not really sure I understand the purpose of this change or how it's inherently making things more DI friendly. What is the goal here?

Reply:
No explicit calls to GetService.  Let DI handle that.

Can provide HttpClient and Logger via DI

Reply:
Under the hood, the ASP.NET Core framework is using IServiceProvider to supply implementations based on the requested types, so it's not really doing anything different in that regard - the main difference is just whether you put all of the dependencies directly in the constructor signature or not. Generally speaking, I was always taught to avoid optional parameters in constructors for production code, and instead have explicit constructors. Using IServiceProvider allows us to keep the constructor small and avoid optional parameters. There's also nothing stopping HttpClient or ILogger being supplied by DI already.

Reply:
That constructor exists in the base class.  

How do you propose someone supplies a custom HttpClient?

Reply:
Use services.AddHttpClient to register a client in Startup with the default IHttpClientFactory implementation and allow IServiceProvider to resolve it (whether directly or indirectly) as per the usual approach?

Reply:
How does that make it into the Adapter?  The only path I see is via constructor.  The original path will result in none (because of a default of null), and eventually, a default one will be created via "new HttpClient".

Reply:
But it's not just HttpClient.  How would they have a custom retry policy or ILogger?

We should at least allow the more flexible version of BotFrameworkHttpAdapter, no?

Reply:
Let me clarify: I think it is fine to pass along an HttpClient object into the base constructor, no objection to that. However, what I'm saying is that I think you can do that by doing something like:

customHttpClient: services.GetService<IHttpClientFactory>().CreateClient()

The main hesitation I have if anything about taking an HttpClient object directly in the CoreBotAdapter constructor is that it's quite dumb, in the sense that it's whatever the default client that's been registered with the factory is. This goes back to my general frustration of making assumptions about the state of the app settings and config and enforcing a particular structure on the developer.

For example, I think it would make sense to allow using a named client so that way developers can register multiple clients with the factory if they so choose, otherwise we can default to one.

Reply:
I buy all that.